### PR TITLE
Initial DatumHash -> Datum index implementation.

### PR DESCRIPTION
**Pull request by raduom**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-chain-index/app/Indexer.hs
Comment (andreabedini) [inline plutus-chain-index/app/Indexer.hs:51]: :-/ I had made this simpler in https://github.com/input-output-hk/cardano-node/pull/3619 but we don't have it yet.

File: plutus-chain-index/src/Cardano/Index/Datum.hs
Comment (andreabedini) [inline plutus-chain-index/src/Cardano/Index/Datum.hs:5]: Why does this start with `Cardano.` and not with `Plutus` like all our packages?

Can this be `Plutus.Index.Datum`? even the executable is `plutus-indexer` (also `index`? `indexer`? pick one :-))

File: plutus-chain-index/app/Indexer.hs
Comment (andreabedini) [inline plutus-chain-index/app/Indexer.hs:73]: is this ever going to happen using `runChainSync` ?

File: plutus-chain-index/app/Indexer.hs
Comment (andreabedini) [inline plutus-chain-index/app/Indexer.hs:101]: We need to be able to pass a different location of the node socket.

File: plutus-chain-index/app/Indexer.hs
Comment (andreabedini) [inline plutus-chain-index/app/Indexer.hs:79]: Can we use `modifyIORef`? (or the strict one `modifyIORef'`)

```
modifyIORef' ixref (Ix.insert . getDatums blk)
```

**Review by andreabedini — CHANGES_REQUESTED** [review]
I assume this is for https://input-output.atlassian.net/browse/PLT-89. Good work.

This PR still uses the chain sync client rather than the streaming one. Hopefully I will get #476 in soon.

Regarding the `hysterical-screams` dependency, we are waiting on a new repo, then it will be merged in.

I left some comments and I have few requests:

1) We need to be able to pass the node socket from the command line, we cannot hard code it
2) I think we should align the module name with the repo and executable, so `Plutus.` not `Cardano.`
3) We need a README
4) How do we query this? Directly from SQLite? can you add a query example in the README?

**Comment by raduom** [comment]
Here are some clarifications regarding `Indexer.hs`. It's not here to stay. The only purpose of this executable is to gather some performance informations, make sure we can sync with the mainnet fully and under what conditions. It will be replaced by the streaming library implementation, once it gets merged.

File: plutus-chain-index/app/Indexer.hs
Comment (raduom) [reply plutus-chain-index/app/Indexer.hs:32]: Yes. This binary is only here to be able to test that we can sync with the main net.

File: plutus-chain-index/app/Indexer.hs
Comment (raduom) [reply plutus-chain-index/app/Indexer.hs:47]: Yes, that is the reason. I could also remove it as it takes only about 15 minutes to get to this point, but I will add a comment.

File: plutus-chain-index/app/Indexer.hs
Comment (raduom) [reply plutus-chain-index/app/Indexer.hs:51]: Yeah. This is just temporary (tm) :)

File: plutus-chain-index/src/Cardano/Index/Datum.hs
Comment (raduom) [reply plutus-chain-index/src/Cardano/Index/Datum.hs:5]: My reasoning is that Indexers are generally more connected to `Cardano`, since that is what we are indexing, the blockchain. For example, `UTXOs` don't have any connection with `Plutus`.

File: plutus-chain-index/app/Indexer.hs
Comment (raduom) [reply plutus-chain-index/app/Indexer.hs:73]: Probably not, but this is just a temporary way of checking that the code can sync with the main net.

File: plutus-chain-index/app/Indexer.hs
Comment (raduom) [reply plutus-chain-index/app/Indexer.hs:101]: Do you still think we need to do this, if we just use the client until `plutus-streaming` gets merged?

File: plutus-chain-index/app/Indexer.hs
Comment (raduom) [reply plutus-chain-index/app/Indexer.hs:79]: I don't think we can, since `Ix.insert` is an effect producing function, but we don't need to worry about concurrency either since nothing is concurrent right now.

Comment (andreabedini) [reply plutus-chain-index/app/Indexer.hs:82]: Oh right, :facepalm:

**Comment by andreabedini** [comment]
> Here are some clarifications regarding `Indexer.hs`. It's not here to stay.

/shrug The things I pointed out are needed for [PLT-89](https://input-output.atlassian.net/browse/PLT-89) to be done. I don't know why you keep saying everything is temporary.

> It will be replaced by the streaming library implementation, once it gets merged.

I just managed to merge it (finally).

**Comment by raduom** [comment]
> 4. How do we query this? Directly from SQLite? can you add a query example in the README?

We don't right now. I am not sure how we could expose this data in a consistent way, but for now we could just query the sqlite db.

File: plutus-chain-index/app/Indexer.hs
Comment (andreabedini) [reply plutus-chain-index/app/Indexer.hs:104]: I am not expecting the user to having to change their node configuration to use this indexer. Yes, I don't think hardcoding the socket path is a good idea. (PS, I just merged #476)

**Comment by andreabedini** [comment]
> We don't right now. I am not sure how we could expose this data in a consistent way, but for now we could just query the sqlite db.

Can you add some documentation where you show an example of how to query this from sqlite?

**Comment by raduom** [comment]
> Can you add some documentation where you show an example of how to query this from sqlite?

Added a comment at the top of the file.

File: plutus-chain-index/src/Cardano/Index/Datum.hs
Comment (koslambrou) [reply plutus-chain-index/src/Cardano/Index/Datum.hs:5]: I agree with @raduom . However, the module name should reflect the package's name. For example, if we call the package `Marconi`, then the module should be `Marconi.Index.Datum`.

File: plutus-chain-index/src/Cardano/Index/Datum.hs
Comment (koslambrou) [inline plutus-chain-index/src/Cardano/Index/Datum.hs:62]: Use `SecurityParam` from Cardano.Api.

File: plutus-chain-index/src/Cardano/Index/Datum.hs
Comment (koslambrou) [inline plutus-chain-index/src/Cardano/Index/Datum.hs:70]: Add haddock comments on the parameters.
Especially for `[Event]`, why do we need this to query the database?

File: plutus-chain-index/app/Indexer.hs
Comment (koslambrou) [inline plutus-chain-index/app/Indexer.hs:74]: For datums, yes. But some indexers will want to index from start (like the utxosByAddress indexer)

**Review by koslambrou — CHANGES_REQUESTED** [review]

File: plutus-chain-index/app/Indexer.hs
Comment (raduom) [reply plutus-chain-index/app/Indexer.hs:74]: I will remove it when I will add the other indexer.

**Comment by koslambrou** [comment]
> > Here are some clarifications regarding `Indexer.hs`. It's not here to stay.
> 
> /shrug The things I pointed out are needed for [PLT-89](https://input-output.atlassian.net/browse/PLT-89) to be done. I don't know why you keep saying everything is temporary.
> 
> > It will be replaced by the streaming library implementation, once it gets merged.
> 
> I just managed to merge it (finally).

Since this PR works by itself, how about we do the integration with `plutus-streaming` in a separate PR?

File: plutus-chain-index/src/Cardano/Index/Datum.hs
Comment (raduom) [reply plutus-chain-index/src/Cardano/Index/Datum.hs]: We need something to query the database plus the information that is currently in memory.

**Review by koslambrou — APPROVED** [review]

File: plutus-chain-index/src/Marconi/Index/Datum.hs
Comment (koslambrou) [inline plutus-chain-index/src/Marconi/Index/Datum.hs:80]: I think the examples got mixed up. `query ix <hash> []` should be for *all* received events, no?

Comment (raduom) [reply plutus-chain-index/src/Marconi/Index/Datum.hs:80]: Nope. It's correct. You may want to drop some of the events <K according to risks that you may want to take. So without any events you get a version of the database that has been settled.

The first example adds all unsettled events (<K) and then runs the query.

**Review by andreabedini — APPROVED** [review]

**Pull request merged**
